libvast/src/logical_pipeline.cpp

```diff
@@ -0,0 +1,421 @@
+  // that schema yet, we instantiate `op` and a schema-specific queue. The
+  // instantiated operator is provided a generator that tries to pull new
+  // elements from the queue, and the input batch we are dispatching is pushed
+  // to this queue. Because this provides additional input to to the
+  // corresponding instance, we iterate over it in order to stream its
+  // results if possible. This does not necessarily already yield results, due
+  // to operators such as `summarize`. Hence, when the input generator is
+  // exhausted, we signal the generators that pull from the queue that they
+  // should stop as well. Finally, we iterate once more over all operator
+  // instances. They will notice that their input is exhausted, which means
+  // that they will become exhausted themselves eventually.
+
+  // We can capture references to `instance_state` due to pointer stability.
+  auto instances = std::unordered_map<type, instance_state>{};
+  auto stop = false;
+
+  /// Dispatches an input batch, creating physical operators on demand.
+  auto dispatch = [&]<class InputBatch>(InputBatch input)
+    -> caf::expected<std::reference_wrapper<instance_state>> {
+    // Empty batches should be handled before calling this.
+    VAST_ASSERT(batch_traits<InputBatch>::size(input) != 0);
+    // For every input batch, we take the following steps:
+    // 1. Find the input schema.
+    auto input_schema = batch_traits<InputBatch>::schema(input);
+    // 2. Try to find an already existing generator, or create a new one
+    // if it doesn't exist yet for the given input schema.
+    auto lookup = instances.find(input_schema);
+    if (lookup == instances.end()) {
+      auto physical_op = op->make_runtime_physical_operator(input_schema, ctrl);
+      if (not physical_op) {
+        return std::move(physical_op.error());
+      }
+      lookup = instances.emplace_hint(lookup, input_schema, instance_state{});
+      auto& instance = lookup->second;
+      instance.physical_op = std::move(*physical_op);
+      instance.gen = make_producer_from_queue<InputBatch>(instance.queue, stop,
+                                                          instance.physical_op);
+      // 3. Push the input element into the buffer.
+      instance.queue.push(std::move(input));
+      instance.it = instance.gen.begin();
+    } else {
+      // 3. Push the input element into the buffer.
+      auto& instance = lookup->second;
+      if (instance.it != instance.gen.end()) {
+        instance.queue.push(std::move(input));
+      }
+    }
+    return std::ref(lookup->second);
+  };
+
+  for (auto&& input : previous) {
+    if (size(input) == 0) {
+      VAST_INFO("inner stalled");
+      co_yield {};
+      continue;
+    }
+    auto instance = std::visit(dispatch, std::move(input));
+    if (!instance) {
+      ctrl->abort(std::move(instance.error()));
+    }
+    // Iterate until the queue is empty.
+    for (auto&& output : exhaust_on_stall(*instance)) {
```

Comment:
An empty queue and stalling is not really equivalent here. This might also depend on the requirements that we place upon implementations. However, it probably works out either way.

Reply:
`exhaust_on_stall` is now `exhaust_on_empty_queue_and_stall`